keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -0,0 +1,1212 @@
+revoked.
+
+Unfortunately that only covers a portion of the behavior that ReferenceGrant
+enables. APIs that make use of ReferenceGrant will need to clearly define how
+the revocation of a ReferenceGrant works. With that said, we expect the
+following guidelines to be rules to apply to **ALL** implementations of the API:
+
+* Deletion of a ReferenceGrant means the granted access is revoked.
+* ReferenceGrant controllers must remove any configuration generated by the
+  granted access as soon as possible (eventual consistence permitting).
+* Some actions that have already been enabled by the ReferenceGrant (such as
+  forwarding requests or persisting data) cannot be undone, but no future
+  actions should be allowed.
+
+### Example Usage
+
+#### Gateway API Cross-Namespace Secret Reference
+
+Authors of APIs can bundle ClusterReferenceGrants with their API definitions
+for any references they expect to be common.
+API's can opt-in to partition references by class names.
+Classes are a normalization of the IngressClass, GatewayClass, StorageClass
+pattern. The ClusterReferenceGrant only needs to extract class names -- no
+knowledge of the specialized IngressClass/GatewayClass/StorageClass API is
+needed.
+References and classes across multiple API versions are supported.
+
+For example, Gateway API could bundle the following ClusterReferenceGrant to 
+authorize same-namespace references from Gateways to Secrets for "tls-serving"
+as well as same-namespace references from Gateways to Secrets and ConfigMaps for "tls-client-validation".
+Gateways are partitioned by GatewayClass. Access to references of a particular
+gatewayClassName should only be given to ClusterReferenceConsumers pertaining to
+that class.
+Proposed fields from GEP-91 are added to the v1 version, but not v1beta1.
+
+```yaml
+kind: ClusterReferenceGrant
+apiVersion: reference.authorization.k8s.io/v1alpha1
+metadata:
+  name: gateways
+from:
+  group: gateway.networking.k8s.io
+  resource: gateways
+versions:
+  - version: v1
+    classPath: ".spec.gatewayClassName"
+    references:
+      - path: "$.spec.listeners[*].tls.certificateRefs[[?(@.group=='' && @.kind=='Secret')].name"
+        to:
+          group: ""
+          resource: secrets
+        for: tls-serving
+      - path: "$.spec.listeners[*].tls.clientValidation.caCertificateRefs[[?(@.group=='' && @.kind=='Secret')].name"
+        to:
+          group: ""
+          resource: secrets
+        for: tls-client-validation
+      - path: "$.spec.listeners[*].tls.clientValidation.caCertificateRefs[[?(@.group=='' && @.kind=='ConfigMap')].name"
+        to:
+          group: ""
+          resource: configmaps
+        for: tls-client-validation
+  - version: v1beta1
+    classPath: ".spec.gatewayClassName"
+    references:
+      - path: "$.spec.listeners[*].tls.certificateRefs[[?(@.group=='' && @.kind=='Secret')].name"
+        # Kubernetes JSONPath doesn't actually support this kind of compound boolean filter expression
```

Comment:
This is more explicit and more annoying than trying to build some sort of duck-typing behavior.
It does make it easier to disambiguate any use of `kind` which is preferable: [API Conventions: kind vs. resource](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#kind-vs-resource)

We still should probably extract namespace here when possible, even if it's just to rule out cross-namespace references. We could access the namespace of the root level object in JSONPath or CEL or something.

Maybe if the path returns a struct of {name, namespace}, we can use it and compare?
Do we force API owners to somehow explicitly declare references are same-namespace?

Do we hold onto cross-namespace references that are declared here as well and try to match those up with any ReferenceGrants from those namespaces?

Presumably, we always ignore {group, resource, kind} in the struct because it must be opt-in.
We can't always know the shape of the API we are descending into.
... unless we can add some sort of safe, sensible matching behavior that automatically checks for the most canonical kinds of object references?

Reply:
There might be APIs that references just a named resource without specifying its `kind` explicitly in the reference. I was thinking that the output of the CEL/JSONPath expression can be something like `kind/ns/name` so the authorizer would know what referenced resource it should look for.
Unless I am missing something and this design already account for such references